File: binary_search/69.py
# 69. Sqrt(x)
# Implement int sqrt(int x).

# Compute and return the square root of x, where x is guaranteed to be a non-negative integer.

# Since the return type is an integer, the decimal digits are truncated and only the integer part of the result is returned.

# BS
class Solution:
    def mySqrt(self, x):
        """
        :type x: int
        :rtype: int
        """
        if x <= 1: return x
        left, right, mid = 1, x, 0
        while left <= right:
            mid = left + (right - left) // 2
            #if mid * mid == x:
            if mid == x / mid: # to avoid overflow
                return mid
            elif mid < x / mid:
                left = mid + 1
                # res = mid # keep the lower int bound in case x's sqrt is not an integer
            else:
                right = mid - 1
        return left - 1
# Newton
class Solution:
    def mySqrt(self, x):
        """
        :type x: int
        :rtype: int
        """
        if x <= 1: return x
        guess = x
        while guess > x / guess:
            guess = (guess + x / guess) // 2
        return int(guess)        

File: binary_search/69.java
public class Solution{
   
        public int mySqrt(int x) {
            int left = 1, right = x;
            int mid;
            if (x < 1) {
                return 0;
            }
            while (left <= right) {
                mid = left + (right - left) / 2;
                if (mid == x / mid) {
                    return mid;
                } else if (mid < x / mid) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
               
            }
            return left-1; //largest integer <= sqrt(x)
        }

    
     public static void main(String []args){
        Solution s = new Solution();
        System.out.println(s.mySqrt(10));
     }
}

class Solution {
    public int mySqrt(int x) {
        if (x < 1) {
            return 0;
        }
        long guess = x; // in case of integer overflow
        while (guess > x / guess) {
            guess = (guess + x / guess) / 2;
        }
        return (int)guess;
    }
}